### PR TITLE
2172 caching grid for randaffine

### LINKED ISSUE #2172: Cache grid info of `RandAffined` transform
**Is your feature request related to a problem? Please describe.**
Feature from internal user:
Hello. I'm working with generative models for 3D scans (size ~200^3). Due to the cost of performing the RandAffine transform on cpu, the gpus are often waiting for the data loader.
`create_grid` takes up a large fraction of the time. By caching the cartesian grid between iterations (and creating copies when applying the transform) we can reduce the cost of RandAffine on cpu by 4 (0.55 to 0.13 sec) and on gpu by 100 (0.45 to 0.0045 sec). This change should allow the client to scale up the inputs when needed.
Caching is not particularly elegant tho. Would a pull request along these lines be acceptable?

The change would be something like adding this method:
```py
def create_cartesian_grid(
        self,
        sp_size: Union[Sequence[float], float]
    ) -> Union[np.ndarray, torch.Tensor]:
        if self.use_cached_grid:
            if self.cached_grid is None or not sp_size == self.cached_sp_size:
                self.cached_sp_size = sp_size
                self.cached_grid = create_grid(spatial_size=sp_size)
                if not torch.is_tensor(self.cached_grid):
                    self.cached_grid = torch.tensor(self.cached_grid)
                if self.device:
                    self.cached_grid = self.cached_grid.to(self.device)
            return self.cached_grid
        else:
            return create_grid(spatial_size=sp_size)
```
In this case, self.use_cached_grid would be given by argument in the constructor and if False the behavior would be same as now.

### PR DESCRIPTION
**Pull request by wyli**
Signed-off-by: Wenqi Li <wenqil@nvidia.com>

Fixes #2172

### Description
adds caching for the identity grid when the spatial size is not dynamic

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/transforms/spatial/array.py [hunk 1]
```diff
@@ -1401,9 +1405,47 @@ def __init__(
...
+        Return a cached or new identity grid depends on the availability.
+
+        Args:
+            spatial_size: non-dynamic spatial size
+        """
+        ndim = len(spatial_size)
+        if spatial_size != fall_back_tuple(spatial_size, [1] * ndim) or spatial_size != fall_back_tuple(
+            spatial_size, [2] * ndim
+        ):
+            raise RuntimeError(f"spatial_size should not by dynamic, got {spatial_size}.")
```
Comment (rijobro) [inline monai/transforms/spatial/array.py:1446]:
```suggestion
            raise RuntimeError(f"spatial_size should not be dynamic, got {spatial_size}.")
```

File: monai/transforms/spatial/array.py [hunk 2]
```diff
@@ -1401,9 +1405,47 @@ def __init__(
...
+        return create_grid(spatial_size=spatial_size) if self._cached_grid is None else self._cached_grid
```
Comment (rijobro) [inline monai/transforms/spatial/array.py:1447]:
```suggestion
        return self._cached_grid or create_grid(spatial_size=spatial_size) if self._cached_grid
```

File: monai/transforms/spatial/dictionary.py [hunk 3]
```diff
@@ -711,6 +712,9 @@ def __init__(
                 Padding mode for outside grid values. Defaults to ``"reflection"``.
                 See also: https://pytorch.org/docs/stable/nn.functional.html#grid-sample
                 It also can be a sequence of string, each element corresponds to a key in ``keys``.
+            cache_grid: whether to cache the identity sampling grid.
+                If the spatial size is not dynamically defined by input image, enabling this option could
```
Comment (rijobro) [inline monai/transforms/spatial/dictionary.py:716]:
```suggestion
                If the spatial size is not dynamically defined be input image, enabling this option could
```

**Review by rijobro — APPROVED** [review]
Thanks!

**Comment by rijobro** [comment]
Perhaps I missed it, but in the unit test, do you ever check that the cached grid is being used? Or perhaps just check that the cached grid is not None.

File: monai/transforms/spatial/array.py (hunk 2 above)
Comment (wyli) [reply monai/transforms/spatial/array.py:1447]: this would give `RuntimeError: Boolean value of Tensor with more than one value is ambiguous`

**Comment by wyli** [comment]
> Perhaps I missed it, but in the unit test, do you ever check that the cached grid is being used? Or perhaps just check that the cached grid is not None.

thanks -- added not None tests

**Pull request merged**